centaur/src/main/scala/centaur/test/metadata/WorkflowFlatMetadata.scala

```diff
@@ -65,30 +63,130 @@ case class WorkflowFlatMetadata(value: Map[String, JsValue]) extends AnyVal {
-        (expected != JsString(o.value.toString)).option(s"expected: $cacheSubstitutions but got: $actual")
+        expected match {
+          case JsBoolean(value) => (value != o.value).option(s"expected: $cacheSubstitutions but got: $actual")
+          case _ => Option(s"metadata $actual is a boolean, but expected to be the same type as $expected")
+        }
       case o: JsArray if stripQuotes(cacheSubstitutions).startsWith("~>") =>
         val stripped = stripQuotes(cacheSubstitutions).stripPrefix("~>")
         val replaced = stripped.replaceAll("\\\\\"", "\"")
-        (replaced != o.toString).option(s"expected: $cacheSubstitutions but got: $actual")
-      case o: JsArray => (expected != JsString(o.toString)).option(s"expected: $cacheSubstitutions but got: $actual")
+        val expectedArray = stringToArray(replaced)
+        expectedArray match {
+          case Success(array) =>
+            (array.elements != o.elements).option(s"expected: $array but got: $actual")
+          case Failure(_) => Option(s"metadata $actual is an array, but expected to be the same type as $replaced")
+        }
+      case o: JsArray =>
+        expected match {
+          case JsArray(elements) =>
+            (elements != o.elements).option(s"expected: $cacheSubstitutions but got: $actual")
+          case _ => Option(s"metadata $actual is an array, but expected to be the same type as $expected")
+        }
       case JsNull => (expected != JsNull).option(s"expected: $cacheSubstitutions but got: $actual")
       case _ => Option(s"expected: $cacheSubstitutions but got: $actual")
     }
 
     matchError.map(s"Metadata mismatch for $key - " + _)
   }
+
+  private def stringToArray(string: String): Try[JsArray] =
+    if (!string.startsWith("[") || !string.endsWith("]")) {
+      Failure(new IllegalArgumentException(s"Metadata $string is not a valid array"))
+    } else {
+      Success(stringToArrayHelper(string))
+    }
+
+  private def stringToArrayHelper(string: String): JsArray = {
+    val brokenUpString = string.substring(1, string.length - 1).split(",")
+    JsArray(brokenUpString.toVector.map { s =>
+      // If this is a multi-dimensional array...
+      if (s.startsWith("[") && s.endsWith("]")) {
+        stringToArrayHelper(s)
+      } else {
+        JsString.apply(s.stripPrefix("\"").stripSuffix("\""))
+      }
+    })
+  }
 }
 
 object WorkflowFlatMetadata {
 
-  def fromConfig(config: Config): ErrorOr[WorkflowFlatMetadata] =
-    config.extract[Map[String, Option[String]]] match {
-      case Result.Success(m) => Valid(WorkflowFlatMetadata(m safeMapValues { _.map(JsString.apply).getOrElse(JsNull) }))
-      case Result.Failure(_) => invalidNel(s"Metadata block can not be converted to a Map: $config")
+  def fromConfig(config: Config): ErrorOr[WorkflowFlatMetadata] = {
+    val values: scala.collection.mutable.Map[String, JsValue] = scala.collection.mutable.Map[String, JsValue]()
+    config.entrySet().stream().forEach { value =>
```

Comment:
As above - any chance spray can help us with this?

Reply:
I don't think spray can help us with this one, by this point we aren't dealing with JSON anymore. I looked around for a function that could do this for us and it doesn't seem like any such function exists.